### ISSUE TITLE
Introduce search history in NvDA Find dialog

### ISSUE THREAD

**Issue by josephsl** [nvaccess/nvda#8482]
Hi,

This is the "issueformation" (combination of the words "issue" and "transformation") of the PR #6484:

### Rationale
Suppose one wishes to find multiple texts from different applications. Currently, in order to find new text in any application that supports cursor manager, users must enter the new text. In case users may wish to find something they've searched before, a search history in the form of a combo box should be provided.

### Steps to reproduce:
1. Open multiple document windows (web browsers, PDF documents, several Word documents and such), all with browse mode turned on.
2. While using one of them, press Control+NVDA+F to open Find dialog.
3. Search for some text and press Enter.
4. Reopen Find dialog (instead of using find next/previous command) and search for another text.
5. Repeat step 4.

### Actual behavior:
Users must enter new text again.

### Expected behavior:
Users can choose items from search history.

#### NVDA Installed/portable/running from source:
Installed

#### NVDA version:
Alpha/future version

#### Windows version:
Windows 10 Version 1803

#### Name and version of other software in use when reproducing the issue:
N/A

#### Other information about your system:
N/A

#### Does the issue still occur after restarting your PC?
Yes

#### Have you tried any other versions of NVDA?
Yes (perpetual as this is a new feature)

### Technical details
One way to implement this is via wx.ComboBox and hosting search history (a list/priority queue) inside Cursor Manager class (the base class, that is). When Find dialog opens, present this combo box with entries coming from search history list, with find next (NvDA+F3)/previous (NVDA+Shift+F3) searching the topmost item.

Special corner case: what if users type something that NVDA knows is already part of search history (the whole text, that is)? If so, bump this text to the top if it isn't.

Another case: suppose there is already an entry in search history and the user types the same entry that only differs in case. If so, the best path might be to compare the texts regardless of case, or if case sensitivity is turned on, just prepend the new variant.

For a code example, see StationPlaylist Studio add-on.

Thanks.

**Comment by amirsol81** [comment]
I also support the implementation of this handy feature.

**Comment by marlon-sousa** [comment]
PR #8921 opened and ready for review.

**Comment by Adriani90** [comment]
I am reopening this issue since the PR has been reverted.
@josephsl, @feerrenrut, @leonardder and others, please when reverting a feature or something that was related to an issue, please don't forget to reopen the issues. Thanks.
Reverting for this one was in PR #9993.

**Comment by levilibegas** [comment]
I wonder if it would be possible to set back Search History, and if the program would let users use it in case it was usable, because they are not using ime for writing. It was very useful, and that could be at least useful for many. Could you please consider bringing it back? Thanks

**Comment by marlon-sousa** [comment]
I have recently moved to another city and changed jobs. Although this is personal info and is not related to the issue itself it explains why I have been unable to loolk iinto this during last months, specially trying to avoid reversal when 2019.2 was released.
I don't think we should prevent ime users from having the feature only because wx widgets (either it or the python port I am yet to figure out) has the unfortunate behavior theyhave whenn pressing esc on an edit combo.
This should be fixed by them or we shall find another way of building the dialog. Unfortunately, I could not figure out myself how to simulate the ime problem, I tried to install japanese language but am lost as how to simulate the right and wrong behaviors using it. Arrows do not read characters and I don't know if the right behavior is really right or not.

I am perfectly opened to hearing technical suggestions to address the issue before folks from wxpython provide a fix meanwhile, but am still without ideas as how to address stuff.

**Comment by levilibegas** [comment]
I wonder if you could consider restoring it the search history, in case it could be usable (for system not using ime), I mean for western languages, at least... thanks

**Comment by nvdaes** [comment]
Hi, I perceived a slightly slower response (more time needed to open the Find dialog) when search history was introduced, using Spanish. What about using a combo box separated from the edit box? We use this in placeMarkers add-on and I think that the responsiveness can be faster. Just an alternative.

**Comment by levilibegas** [comment]
Do ou mean the history search is back? I am using last alfa and I cannot see it. THANKS.


Enviado desde mi iPhone

El 23 oct 2019, a las 5:55, nvdaes <notifications@github.com> escribió:

﻿
Hi, I perceived a slightly slower response (more time needed to open the Find dialog) when search history was introduced, using Spanish. What about using a combo box separated from the edit box? We use this in placeMarkers add-on and I think that the responsiveness can be faster. Just an alternative.

**Comment by marlon-sousa** [comment]
If project owners wish they can do it.


As I said, I don't feel comfortable doing that because if we do it the 
pressure to restore the thing to everyone will low and we will end up 
concluding that once ime is challenging then we can postpone the feature 
on and on to them.

This opens the door for exclusion among ourselves and this is dangerous.


Being the creator of the PR I for sure need the feature but prefer to 
wait until we have a global working solution. if I had feedback from 
Asian languages users during the many months this was available I may be 
would had time to fix that. Now it's late and releases have gone. We 
took more than one year to deploy such a simple feature and that is 
frustrating. Anyways I will investigate further possibilities. In the 
meantime I am still hearing suggestions.

**Comment by levilibegas** [comment]
>i am not a proramer, so I can be of not help apart from testing. I can ty some chinese friends so they woud test? what should they test? thanks, muito obrigado!!


Enviado desde mi iPhone

El 23 oct 2019, a las 12:13, Marlon Brandão de Sousa <notifications@github.com> escribió:

﻿If project owners wish they can do it.


As I said, I don't feel comfortable doing that because if we do it the 
pressure to restore the thing to everyone will low and we will end up 
concluding that once ime is challenging then we can postpone the feature 
on and on to them.

This opens the door for exclusion among ourselves and this is dangerous.


Being the creator of the PR I for sure need the feature but prefer to 
wait until we have a global working solution. if I had feedback from 
Asian languages users during the many months this was available I may be 
would had time to fix that. Now it's late and releases have gone. We 
took more than one year to deploy such a simple feature and that is 
frustrating. Anyways I will investigate further possibilities. In the 
meantime I am still hearing suggestions.

**Comment by nvdaes** [comment]
Hi, if your answer is for my comment, quoted in your reply, I was 
talking about search history previously implemented and removed, and the 
alternative used in placeMarkers add-on to save, if the corresponding 
checkbox is checked, previous strings searched, but in this case it's 
used for specific documents and saved permanently in text files used 
when the corresponding page is reloaded. Here we use a conventional 
combo box, and I was suggesting this as an alternative.
You can see documentation about placeMarkers at
https://addons.nvda-project.org/addons/placeMarkers.en.html
I feel that this dialog is more responsive comparing it with the removed 
dialog used for search history, even in languages like Spanish.

**Comment by marlon-sousa** [comment]
Hello,

I have been able to overcome the ime issue. Now, escape key will not close the dialog if IME is oppenned.

I can submit another pull request or open the old one again, but I wanted to submit this for testing by someone who is a fluent ime user, as I have done all I could to test it but may be something is missing.

Are any of you whiling to test it?

If so you can either clone my fork and use the previousTermsInSearchDialog branch or I can try to build a running copy out of it and make it available for download.

**Comment by josephsl** [comment]
Hi, I’ll take a look at it – I also suggest posting a pull request to wxPython repo if this has something to do with it. Thanks.

**Comment by marlon-sousa** [comment]
Hello,


I didn't play with wxpython, not at this time.


What I did is I exposed the ime status through a module variable in 
module NVDAHelper. This will benefit also addons that in the future need 
to check ime status as a kind of bonus.


The find dialog then (at the dialog level) checks if ime is opened. If 
so then the escape key is not processed, although it is processed at the 
edit combo level, thus preventing the dialog from closing.


If IME is not active then the default behavior is used.

**Comment by josephsl** [comment]
Hi, somehow, it fails for Korean. When I type Korean text and I press Escape, search history dialog closes. Thanks.

**Comment by marlon-sousa** [comment]
I have tested with Japanese keyboard. Can you please describe me carefully the procedures you used to type Korean. have you cloned my project And selectedem Selected de leite apropriate Branchbranch? It’s not merged into Master. Obrigado,
Marlon

**Comment by josephsl** [comment]
Hi, after installing Korean IME, switch to Korean (Windows+SPACE), then press right Alt key to switch to “native input” and type Korean characters (any character will do for this test). As one types Korean characters, a square composition window will appear, used to complete actual Korean characters and words. Thanks.

**Comment by Qchristensen** [comment]
Has anyone looked at this issue recently?  It is still being requested.

**Comment by nvdaes** [comment]
Íf a combo box independent to the edit box, set as a different control, is accepted, we used it in placeMarkers addon, but this is saved for different documents based on their title and other parameters (as a feature named specific search). In case it"s useful let me know and I"ll have a look in the next weekend. PlaceMarkers is in the community add-ons website and specific search can be activated with control shift NVDA f. The add-on repo is in github bar nvdaes placeMarkers. (I am commenting from mobile, so it"s more difficult to search and type urls).

**Comment by marlon-sousa** [comment]
Hello,

Due to several reazons, the associated prs weren't and don't seen set to be merged intto NVDA very soon. I took this out of the screen reader itself and put this very same implementation on an addom I myself and many others have been  using for a long time now without any issues. Refer to https://github.com/marlon-sousa/EnhancedFindDialog and enjoy!

**Comment by CyrilleB79** [comment]
It has also been requested recently in the French community.

Thanks for the add-on link; I did not know of its existence. It is out of  scope of this ticket, but I recommend you to submit it to the add-on community in order to have it available on the add-on community website.

**Comment by LeonarddeR** [comment]
This was reverted by https://github.com/nvaccess/nvda/pull/9993

I recently stumbled upon this again because I was at some point heavily frustrated by the lack of a find history feature 😜
@marlon-sousa and/or @josephsl, could you please:

1. Summarize the reason why this was reverted, along with steps to reproduce, preferably against the enhanced find dialog add-on
2. Elaborate on what the status is currently, i.e. is there any chance we can work around this without a fix in wx? I think it might be possible that this is a bug in win32.

**Comment by marlon-sousa** [comment]
Yes.

This was about to be merged when some people, from Japan as far as I can remember, reported that the arrows used to select through search history were some how interfering with ime output.

I had no access to ime input,  nor had I a clear, structured step by step simulation error report, so that I could reproduce and suggest work arounds.

As I had complains but not support and I needed the feature, I have chosen to follow the addon path, so that everyone (including me) could bennefite from the feature.

I still think that NVDA deserves a search history, because this can greatly enhance productivity of people who work using a web based system.

At the workplace, it is often not easy to install addons.

But I cannot move forward if I have not the needed support, specially because this is way different than my reality.

**Comment by marlon-sousa** [comment]
Don't waste your time testing, the code in the addon is not ime ready.

If you can somehow say me what I need to do to simulate it, I can try to provide a fix.

I remember asking back them how to detect whether ime was active, as I haven't found a reliable system way and I didn't want to include code looking at opened windows or such.

I remember asking them what would be a functional solution, because not being a ime user I don't know how could I make it work (e.e if arrows are used to select characters, then how could we select search entries)? How does this feature behave in JAWS which has search history cinse at least version 3.7 as far as I can remember?

Without adequate support from the community, there is not much I can do about this.

**Comment by cary-rowen** [comment]
I just want to say that pressing ESC when an IME candidate exists will cause the dialog to be closed, which seems to be the behavior of Windows. At least I have always faced this problem.

Of course, this is really annoying, another example is the add-on store, which renders as a dialog, and while I don't know why such a design is used, it's true that I can't press ESC when the candidate appears.

**Comment by marlon-sousa** [comment]
My position on that is simple:
I won't be begging any more to introduce any feature NVDA itself should desire to have, specially things as simple as this one, that all screen readers just have, and are common sense.

I fortunately (as of now) still have the freedom to install addons, so my own productivity is, in this matters, guaranteed due to work, by the way, done by me with the EnhancedFindDialog addon.

I am sorry for those who can't install addons or who need something like this but are not technical enough to know how addons work. Most part of people working with web based systems are not programmers or screen reader power users.

If things advance to a situation where I can't no longer install addons at the workplace, I will have to figure out a way for me.

As for this issue here, whenever someone (who prevented the feature to be merged) offer a solution on what should be done, I will (if I am still available) be more than happy to code it.

The EnhancedFindDialog addon will continue to grow, offering in the future regexp based find operations and permanent search history if users desire.

As for muy trials to help on NVDA, I can only say I am extremely sorry.

**Comment by levilibegas** [comment]
Could it be doable to include it and let know in the users guide the difficulty for some imes? Many users would certainly love to have it embedded into the code while things are arranged for some ime..

**Comment by cary-rowen** [comment]
@josephsl 

I'm surprised!

If #9954 is considered a reasonable request to restore this PR, then the add-on store should also be restored. Because at least Chinese IME cannot use Escape in the add-on store.

**Comment by CyrilleB79** [comment]
@cary-rowen wrote:
> [@josephsl](https://github.com/josephsl)
> 
> I'm surprised!
> 
> If [#9954](https://github.com/nvaccess/nvda/issues/9954) is considered a reasonable request to restore this PR, then the add-on store should also be restored. Because at least Chinese IME cannot use Escape in the add-on store.

I just read this comment. Is this issue logged somewhere? If not can you do so?

**Comment by cary-rowen** [comment]
This has been a fact of life for as long as I can remember, to the point where the behavior described below has become almost a de facto standard.
This seems to have changed with the new version of Microsoft Pinyin IME in Windows 11, which is a recent development.

**Steps to reproduce:**
1.  Install Microsoft Pinyin or the more popular Sogou Pinyin IME in China on Windows 10.
2.  In any input control within NVDA (e.g., the filter in the Input Gestures dialog, the search box in the Add-on Store, the find dialog in browse mode, etc.), type something to bring up the IME candidate list. Then, press ESC to cancel the candidate list instead of pressing Space to select a candidate.

**Actual behavior:**
The dialog box containing the input control closes.

**Expected behavior:**
The candidate list is canceled, but the dialog box does not close.

This expected behavior is now what happens with Microsoft Pinyin on Windows 11.

For users in the Chinese community who have grown accustomed to the "unexpected" behavior, this was never considered a problem; we even saw it as a kind of standard. That's why I am still puzzled and very confused that the find dialog history feature was reverted because of this issue. People have been asking about this feature for many years, but NV Access hasn't provided any further guidance.

The original submitter is also watching this closely, and I believe they are awaiting a decision from NV Access.

**Comment by CyrilleB79** [comment]
OK thanks @cary-rowen for this feedback; sorry, I had missed that.

So to summarize:
* Chinese people are used to the following behaviour: if the IME is opened from an editable combo-box and `escape` is pressed there, `escape` is passed to the initial dialog, which closes, if it is implemented to do so.
  * Is it only in NVDA GUI or in any editable combo-box?
* the new version of Microsoft Pinyin IME in Windows 11 does not show this behaviour anymore
  * From what you know, do some IME still show this behaviour? If yes, are they still supported for a long time? (e.g. if it is on Windows 10 only, the end of support is quite soon)

Is it correct?

**Comment by hwf1324** [comment]
> * Is it only in NVDA GUI or in any editable combo-box?

This is an issue with WX Dialog.

However, it may be possible to patch this issue in the NVDA GUI.

If there is a way to determine whether the current input state is IME input mode, we can check whether the input state is IME input mode in the `wx.EVT_CHAR_HOOK` event to decide whether to pass the `ESC` key to the dialog box.

https://github.com/nvaccess/nvda/blob/78f789d05a2a8e009b9f606efc2297f68865c227/source/gui/settingsDialogs.py#L268-L280

> * From what you know, do some IME still show this behaviour? If yes, are they still supported for a long time? (e.g. if it is on Windows 10 only, the end of support is quite soon)

We cannot guarantee that all IMEs will intercept keyboard events passed to other programs.

**Comment by ramoncorominas** [comment]
> (e.g. if it is on Windows 10 only, the end of support is quite soon)

Just a side note about Windows 10 end of support:

Microsoft Extends Deadline For Windows 10 Users To 2026
https://www.forbes.com/sites/zakdoffman/2025/06/24/microsofts-free-upgrade-deadline-for-400-million-windows-users/

**Comment by josephsl** [comment]
Hi,

Windows 10 end of support: the extended security updates (ESU) requires specific conditions to be met (using a Microsof Account, for example) in order for consumers to receive twelve months of security updates. ESUs do not include new features and changes - only security bugs are fixed.

As for IME detection, one potential way is looking at GUI thread and see which language/IME input mode is set to. Keep in mind that:

1. Not all non-Latin input modes have IME.
2. Not all IME modules/interfaces work the same.
3. Not all input modes might be exposed by Windows.
4. IME behavior differs across Windows 10/11 releases.
5. Hardware keyboard is not the only way to enter text (think touch keyboard, braille, etc.).

As for Korean input, when I switch to Korean while add-on store is open and type something, composition window opens; when Escape is pressed, add-on store closes. If it is indeed something going on with wx, then we do need to discuss this issue with wxWidgets/wxPython folks as it can impact potentially any wxWidgets/wxPython application.

Thanks.

**Comment by cary-rowen** [comment]
Windows 10 stop support or extended support, I think both are trivial for this issue. Nothing should be blocked; if something is blocked, then all existing dialogs need to be fixed to resolve the problem. I believe we worry too much about many things.
If this is really a serious problem (including non-hardware keyboard input), no one has reported it for years. It's not only reproducible in the add-on store but also can be reproduced in the input gesture dialog, which has existed for over ten years.

**Comment by cary-rowen** [comment]
Why do people report issue only after introducing the enhanced search dialog, and why is the preferred solution to roll back the newly added feature (treating it as a regression) rather than fixing the more widespread issue that already existed?

**Comment by ramoncorominas** [comment]
> Windows 10 end of support: the extended security updates (ESU) requires specific conditions to be met (using a Microsof Account, for example) in order for consumers to receive twelve months of security updates. ESUs do not include new features and changes - only security bugs are fixed.

As far as I know, there are no new features since at least January, and I think it is very unlikely that there will be more new features before October, so in that sense we could say the end of support has been already reached, but we are still considering October as the official date and, if I understand it well, linking NVDA support to that date. For the ESU, one of the possible conditions is to pay $30, so I guess many organizations will choose to pay that instead of buying new computers just to support Windows 11.

**Comment by josephsl** [comment]
Hi,

We had new features and changes roll out since January, including Copilot (well, for that one, it is really an app/portal to Bing Search and powered by parts of ChatGPT) and Windows Backup. For organizations, ESU year 1 costs 61 United States dollars.

The NVDA project does not tie NVDA support to Windows end of life dates - the most recent example is Windows 8.0 which was supported by Microsoft and NVDA project until January 2016 and April 2024, respectively (due to Python). I expect Windows 10 to be a supported scenario until at least 2027 (CPython (the actual Python runtime we use) supports oldest server release under long-term support, and for practical reasons, the latest server release that is tied to Windows 10 is Server 2019 which goes out of extended support in January 2029 )mainstream support has ended last year); Server 2022 is a special case in that it includes portions of Windows 11 technologies on top of Windows 10 interface).

As for folks finding errors when enhanced find feature 1.0 was part of alpha and not discussing other IME interaction issues, I think the visibility of the find dialog feature, along with being accustomed to the IME interaction issue over the years could have contributed to that (sometimes our actions and mindset are guided by the technology interface we use over time). Further, it could be possible that some people may have developed workarounds that were in place while a specialized NVDA version (Japanese, for example) or add-ons were in use and was not widespread.

Thanks.

**Comment by LeonarddeR** [comment]
I really want to see this landing in core, so I have the following suggestion and also prototyped it.
It is based on the work by @marlon-sousa in #8921 and the sentence nav add-on, with one major addition: you can enable/disable search history in browse mode settings. Enabling it (default, feature flag) gives you the combo box, disabling it gives you the current edit box. This ensures we have best of two worlds. The people suffering from the IME issue can disable the flag.

Note that Remote connection dialog also has a combobox for history, so it suffers from the same problem. Same applies to add-on store as pointed out by @josephsl . This has never been a reason to delay these features. So may be we should crack this nut and just decide to introduce history regardless. People who have privacy concerns can still disable it and then they have working IME as a bonus.

**Comment by cary-rowen** [comment]
Hi @LeonarddeR 

I would vote for making the find history a default behavior rather than adding a toggle option, for the following reasons:

1. The Chinese IME issue mentioned above is independent of whether we add a search history feature. Even with the current implementation (where history is not preserved), pressing Esc while the IME candidate is open still closes the entire dialog. Therefore, adding a toggle option doesn't solve this upstream defect.
2. We already have precedents for retaining input history. For instance, the Remote Connection dialog has a host history combo box, and it seems no one has ever requested an option to turn it off.
3. Regarding privacy concerns, I'm not entirely convinced. The find dialog is typically used for searching general keywords—are there really significant risks of sensitive information leakage?

Finally, thank you so much @LeonarddeR for taking this on, it's great.

**Comment by CyrilleB79** [comment]
Re privacy, I do not think we have such setting in editors (e.g. notepad++).
So to me, if the history is cleared across sessions, there is no point in adding a privacy setting.

**Comment by marlon-sousa** [comment]
Hello,

Even though I am still trying to find time for my own addon updates, I have researched the search history retention.

Here are my findings:

1. For me, retaining history between sessions must be a opt in (checkbox in nvda settings).
2. Upon acceptance, a quick text explaining privacy implications must be shown (more like this in a while).
3. If accepted, search history would be retained, and this needs to be either by profile or global.

### Protecting search history

User must choose protection level if they choose to enable the historuy retention:

level 1 creaets a key, stores it on a file in %appdata%\nvda and uses this key to decrypt sqlite file on nvda load, encrypt back the updated version when it quits.
level 2 creates via DPAPI an encryption key using CRYPT_PROTECT_PROMPT_ON_UNPROTECT, encrypts the key and saves it. This will be decripted on nvda load, and user will have to authorize it everytime nvda runs.

These options should be explained, and likely profiles would get ;evel 1 by default, which should be enough (or level 1 writes the key to registry, this would work the same way), and some profiles would opt for level 2, receiving then the authorization dialog when activated by the first time on a new session.

### Profiles

We must allow history per profile because many people working with web systems would prefer to set a work profile that they activate only when they are working with that system, otherwise their history would be cluthered.

For me, either level 1 or 2 are mandatory. Still, it should be explained that othr addons can maliciously query the search history, because they run on nvda, and nothing really can be done about it.

I am extending regexp searches and making other improvements in the addon (likely) in next weeks.

I am fully in to porting the addon to nvda core, once nvaccess guarantes that I wouldn't be wasting my time like it happened before.

**Comment by LeonarddeR** [comment]
@cary-rowen 

> I would vote for making the find history a default behavior rather than adding a toggle option...

It will be the default behavior, the toggle can toggle it off though.

>     1. The Chinese IME issue mentioned above is independent of whether we add a search history feature. Even with the current implementation (where history is not preserved), pressing Esc while the IME candidate is open still closes the entire dialog. Therefore, adding a toggle option doesn't solve this upstream defect.

Wow, so in fact for Chinese this problem is already present? I must have missed that and now I'm puzzled why this was pushed back. Is that related to Korean may be? All this IME stuff is abacadabra for me as a user nowing only the Latin, Greek and Hebrew scripts.
2.>      
>     2. We already have precedents for retaining input history. For instance, the Remote Connection dialog has a host history combo box, and it seems no one has ever requested an option to turn it off.

Fair point.

>     3. Regarding privacy concerns, I'm not entirely convinced. The find dialog is typically used for searching general keywords—are there really significant risks of sensitive information leakage?

Not really to me, may be for others though.

@marlon-sousa Your suggestions are valuable but lets leave that a follow up.

**Comment by LeonarddeR** [comment]
@josephsl Could you please re-elaborate on the current state for Korean?

**Comment by marlon-sousa** [comment]
> [@cary-rowen](https://github.com/cary-rowen)
> 
> > I would vote for making the find history a default behavior rather than adding a toggle option...
> 
> It will be the default behavior, the toggle can toggle it off though.
> 
> > ```
> > 1. The Chinese IME issue mentioned above is independent of whether we add a search history feature. Even with the current implementation (where history is not preserved), pressing Esc while the IME candidate is open still closes the entire dialog. Therefore, adding a toggle option doesn't solve this upstream defect.
> > ```
> 
> Wow, so in fact for Chinese this problem is already present? I must have missed that and now I'm puzzled why this was pushed back. Is that related to Korean may be? All this IME stuff is abacadabra for me as a user nowing only the Latin, Greek and Hebrew scripts

So it was for me. As far as I remember back at that time the Japanese community raised a ticket which removed the pr already in beta, on a very strange, very hard to explain act from those who administered that project.

We still have no guarantees though that something like what I proposed would be accepted as of now, even though the rest of the whole project lives and has always lived with the same characteristics, without a single issue.

 2.>
> 
> > ```
> > 2. We already have precedents for retaining input history. For instance, the Remote Connection dialog has a host history combo box, and it seems no one has ever requested an option to turn it off.
> > ```
> 
> Fair point.
> 
> > ```
> > 3. Regarding privacy concerns, I'm not entirely convinced. The find dialog is typically used for searching general keywords—are there really significant risks of sensitive information leakage?
> > ```
> 
> Not really to me, may be for others though.
> 
> [@marlon-sousa](https://github.com/marlon-sousa) Your suggestions are valuable but lets leave that a follow up.

Your point of view, not mine: if it wasn't important by now I wouldn't suggest it here, and the reason why the addon doesn't implement it already is that this is important. We should at least implement level 1, making the history encripted so stupid software made by any basic llm prompter cannot find the histor opened and spread it. If we protect it some how by saving a key in registry and using it to open / close the history, we have a first line defense: almost no efort, big wins.

**Comment by cary-rowen** [comment]
Introducing encryption just for this feature really feels like overengineering to me. It's not about saying privacy isn't important, but for something like find history, it's just overkill. Look at Microsoft Office, Notepad, or even the Windows Run dialog—they all have search or input history, and none of them require encryption. I’ve never even heard of anyone demanding encryption for those things. In other words, if someone can already access your data so easily, what would be the point of this encryption anyway?

**Comment by marlon-sousa** [comment]
Hello, I didn’t research where Windows stores start menu search history, or run dialogue history. But, even if this is stored somewhere opened, the main reason people are using persisted. Search history is likely because they work in a not so accessible, web system, professionally, where they have to search for terms to enable quick navigation. in this situation, they might search for some sensitive information, such as customer name, or other stuff, that I wouldn’t like to expose in a very easy manner. Level one cryptography would be transparent for the final user. if you can show me where Windows stores these history, searches, as well as notepad, I would appreciate as a matter of curiosity.Enviado do meu iPhoneEm 10 de jul. de 2026, à(s) 22:11, wencong ***@***.***> escreveu:﻿cary-rowen left a comment (nvaccess/nvda#8482)
Introducing encryption just for this feature really feels like overengineering to me. It's not about saying privacy isn't important, but for something like find history, it's just overkill. Look at Microsoft Office, Notepad, or even the Windows Run dialog—they all have search or input history, and none of them require encryption. I’ve never even heard of anyone demanding encryption for those things. In other words, if someone can already access your data so easily, what would be the point of this encryption anyway?

**Comment by CyrilleB79** [comment]
The usual way to manage history in such controls is usually, when there is one, to give users the possibility just to remove existing history, either globally, or item by item, e.g. pressing del key or tabbing on a remove button as in browsers search bar.
But my opinion remains that such history do not need to be protected, as it probably isn't in any mainstream software; the number of memorized items sould probably be limited to 10 so that older searches automatically disappear from the history after a while.

**Comment by CyrilleB79** [comment]
Re the history why this feature could not remain in NVDA when it had been introduced 7 years ago, IIRC @marlon-sousa, nor any other available dev, was not able to test with IME; this has probably caused difficulty to estimate how impacting this issue was, and to separate the causes between the new dev from what pre-existed in the combo-box behavior. NV Access has probably estimated the issue was bigger than it was actually.

**Comment by josephsl** [comment]
Hi,

1. On privacy: if search history terms are not encrypted at runtime, there is a possibility that an add-on can mess with them (remember that, in Python, you can override anything at runtime). Further, since NVDA is open-source, anyone can figure out how search terms are encrypted and come up with a decryption engine (or perhaps a stronger encryptor). I vote for a volatile (session-specific) search history first (needed in secure mode, for example), and then consider encryption and/or storage after analyzing user reactions.
2. IME: in 2026, seems you can use Korean IME with NVDA's find dialog - pressing Escape while Chinese character suggestions appear (type Korean letters, press Right Control key, and scroll through Chinese character (Hanja) suggestions) does not close the find dialog anymore. This may have to do more with wxWidgets/wxPython.
3. Back when this feature was under active development, using IME with NVDA find dialog was a big concern.
4. Default behavior: I vote for search history to be the default (and only) behavior going forward to reduce user confusion (by the way, find dialog in NVDA is defined in cursor manager, and introducing search history is a good opportunity to modernize that dialog and add type hints).

Thanks.

**Comment by marlon-sousa** [comment]
Not discussing over that issue in the past, my points were provided already.

As for search history, if we want to keep history micely opened, we at least need to give users the opportunity to disable that if they don't feel confortable with the approach.

I realise our hability to see users needs and measure complexity versus wins didn't improve much cinse when this feature was first tried, in what I consider to be a huge error anyways, at least from my part: stress on me, no feature,  no enhancements,  no mindset change.

I will keep addon development, which I should get right with freedon like I have been doing (though with each time less free time), where storage will deserve the options users are most confortable with, in control of their data if and when they need.

In general, if you (very unlikely) need my help on prs ping me and I will try my best.

**Comment by LeonarddeR** [comment]
@marlon-sousa wrote:
> In general, if you (very unlikely) need my help on prs ping me and I will try my best.

Thanks for your offer.

I think for the first implementation, we'll stick to history per NVDA session. If someone wants to drop his history for whatever reason, he can restart NVDA. If someone wants to preserve history for whatever reason, we can consider preserving history in a follow up.
